Fix parquet row group prefetching and erasing

marin-ma opened the pull request:
Summary: This PR addresses two main issues related to row group management:

Details:
1. Row group prefetching issue. The existing condition to check if a row group has been loaded before is incorrect, which prevents the row group from being prefetched as configured.

2. First row group never get erased. Due to a bug, the first buffered row group never gets erased after loading subsequent row groups.

Fixes: #6115

marin-ma commented:
@majetideepak I'm wondering if this is also a bug? Seems like `rowGroupIds[0]` will never get erased from `inputs_`

https://github.com/facebookincubator/velox/blob/fcd334c97c2aff6a311702948343e93e80297976/velox/dwio/parquet/reader/ParquetReader.cpp#L594-L596

majetideepak commented on velox/dwio/parquet/tests/reader/ParquetReaderTest.cpp:
```diff
@@ -24,6 +24,8 @@ using namespace facebook::velox::dwio::common;
 using namespace facebook::velox::dwio::parquet;
 using namespace facebook::velox::parquet;
 
+DECLARE_int32(parquet_prefetch_rowgroups);
```
We should make this a parquet reader option.
Flags are discouraged.

majetideepak commented:
> @majetideepak I'm wondering if this is also a bug? Seems like `rowGroupIds[0]` will never get erased from `inputs_`

@marin-ma You are right. The check should be `if (currentGroup >= 1)`

majetideepak commented on velox/dwio/parquet/tests/reader/ParquetReaderTest.cpp:
```diff
@@ -595,3 +597,51 @@ TEST_F(ParquetReaderTest, parseLongTagged) {
+
+  // Advance to the second row group.
+  parquetRowReader->nextRowNumber();
+  // Check the next row group has been prefetched.
```
We can probably add `EXPECT_FALSE(parquetRowReader->isRowGroupBuffered(0));` here after fixing the check

marin-ma commented:
@majetideepak Also updated issue/PR description. PTAL. Thanks!

majetideepak commented on velox/docs/develop/TpchBenchmark.rst:
```diff
@@ -115,7 +115,6 @@ Top Optimization Recommendations
    "num_splits_per_file", "Row Group Size of Data", "Row Group Size of Data"
    "max_coalesce_bytes", "Minimum of 90MB", "Minimum of 90MB"
    "max_coalesce_distance_bytes", "Workload dependent**", "Workload dependent**"
-   "parquet_prefetch_rowgroups", "Best is 1", "Best is 1"
```
Let's keep this and add a flag `parquet_prefetch_rowgroups` to TpchBenchmark.cpp.
We can then pass this via the connector properties just like `max_coalesced_bytes`
Also please fix typo here `max_coalesce_bytes` -> `max-coalesced-bytes`
`max_coalesce_distance_bytes` -> `max_coalesced_distance_bytes`

majetideepak reviewed (commented):
@marin-ma Looks great! Just one comment.

majetideepak reviewed (approved):
@marin-ma thanks! Can you rebase with the main branch and resolve the conflicts?

majetideepak commented:
@Yuhta can you take a look?

jinchengchenghh commented on velox/dwio/common/Options.h:
```diff
@@ -462,6 +465,14 @@ class ReaderOptions {
   }
 
+  /**
+   * Modify the number of row groups to prefetch.
```
Please fix the code style as https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md#comments

Here we should use //

jinchengchenghh commented on velox/dwio/parquet/tests/reader/ParquetReaderTest.cpp:
```diff
@@ -630,3 +630,44 @@ TEST_F(ParquetReaderTest, preloadSmallFile) {
+  const int numRowGroups = 4;
+
+  ReaderOptions readerOptions{defaultPool.get()};
+  // Disable preload of file
```
Add the dot in the end of comment

marin-ma commented:
@Yuhta Could you please take a look? Seems like the CI failure is not related to this PR.

marin-ma commented:
@majetideepak @Yuhta Could you please take a look? Can this one get merged?

marin-ma commented:
@Yuhta Could you please take a look?

marin-ma commented:
@mbasmanova Could you take a look?

mbasmanova commented on velox/dwio/parquet/reader/ParquetReader.h:
```diff
@@ -64,6 +64,8 @@ class ParquetRowReader : public dwio::common::RowReader {
     return true;
   }
 
+  bool isRowGroupBuffered(int32_t rowGroupIndex) const;
```
It would be nice to document this API.

mbasmanova commented on velox/dwio/parquet/tests/reader/ParquetReaderTest.cpp:
```diff
@@ -630,3 +630,44 @@ TEST_F(ParquetReaderTest, preloadSmallFile) {
+  ReaderOptions readerOptions{defaultPool.get()};
+  // Disable preload of file.
+  readerOptions.setFilePreloadThreshold(0);
+  readerOptions.setPrefetchRowGroups(1);
```
Should we test some other values as well? E.g. 2 (less than total), 4 (exactly as total), 10 (more than total)?

mbasmanova reviewed (commented):
@marin-ma Looks good to me % a comment about testing. It would be nice to update PR description as it is hard to read and understand. Let's wait for @Yuhta to do a final pass and merge.

majetideepak commented:
@marin-ma Can you share an update on wrapping this up? Thanks.

marin-ma commented:
> @marin-ma Can you share an update on wrapping this up? Thanks.

@majetideepak Addressed comments and rebased. Still waiting for @Yuhta's final review. 

@Yuhta Could you help to review? Thanks!

majetideepak commented:
@Yuhta, @mbasmanova any further feedback? Thanks.

Yohahaha commented on velox/dwio/parquet/reader/ParquetReader.cpp:
```diff
@@ -578,19 +574,18 @@ void ReaderBase::scheduleRowGroups(
     if (nextGroup) {
-      if (inputs_.count(nextGroup) != 0) {
-        inputs_[nextGroup] = reader.loadRowGroup(thisGroup, input_);
+      if (inputs_.count(nextGroup) == 0) {
```
could we use `isRowGroupBuffered` to replace this check?

Yohahaha commented on velox/dwio/parquet/tests/reader/ParquetReaderTest.cpp:
```diff
@@ -632,3 +632,60 @@ TEST_F(ParquetReaderTest, preloadSmallFile) {
+
+  facebook::velox::dwio::common::ReaderOptions readerOptions{defaultPool.get()};
+  // Disable preload of file.
+  readerOptions.setFilePreloadThreshold(0);
```
should we also set directorySizeGuess with a small size? such as 1024, 
 kDefaultDirectorySizeGuess is 1MB.

```c++
  bool preloadFile =
      fileLength_ <= std::max(filePreloadThreshold_, directorySizeGuess_);
```

majetideepak commented:
@Yuhta Can you please review this PR? Thanks.
